src/hit/api/evaluator/explicitdepthfinder.cpp

```diff
@@ -0,0 +1,147 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "explicitdepthfinder.h"
+
+#include <glog/logging.h>
+
+#include "../../common.h"
+
+using namespace std;
+
+namespace hit {
+
+    CKKSCiphertext ExplicitDepthFinder::encrypt(const vector<double> &) {
+        LOG_AND_THROW_STREAM("ExplicitDepthFinder does not define encrypt() with an implicit level");
+    }
+
+    CKKSCiphertext ExplicitDepthFinder::encrypt(const vector<double> &, int level) {
+        if (level < 0) {
+            LOG_AND_THROW_STREAM("Explicit encryption level must be non-negative, got " << level);
+        }
+
+        max_contiguous_depth = max(max_contiguous_depth, level);
+
+        CKKSCiphertext destination;
+        destination.he_level_ = level;
+        // Using a default num_slots_ is potentially problematic if the depth of
+        // the function depends on the number of slots. This seems like an unusual
+        // situation, so its doesn't seem worth fixing.
+        destination.num_slots_ = num_slots_;
+        destination.initialized = true;
+
+        return destination;
+    }
+
+    // print some debug info
+    void ExplicitDepthFinder::print_stats(  // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+        const CKKSCiphertext &ct) {         // NOLINT(readability-convert-member-functions-to-static)
+        VLOG(VLOG_EVAL) << "    + Explicit Level: " << ct.he_level();
+    }
+
+    int ExplicitDepthFinder::num_slots() const {
+        return num_slots_;
+    }
+
+    void ExplicitDepthFinder::set_explicit_post_bootstrap_depth(CKKSCiphertext &ct1, const CKKSCiphertext &ct2) {
+        if (ct1.bootstrapped() != ct2.bootstrapped()) {
+            // levels will not be aligned.
+            const CKKSCiphertext &bootstrapped_ct = ct1.bootstrapped() ? ct1 : ct2;
+            const CKKSCiphertext &fresh_ct = ct1.bootstrapped() ? ct2 : ct1;
+
+            int explicit_bootstrap_lvl = fresh_ct.he_level() - bootstrapped_ct.he_level();
+            if (explicit_post_bootstrap_depth_ < 0) {
+                // we have not yet set the explicit_post_bootstrap_depth_
+                if (explicit_bootstrap_lvl >= 0) {
+                    explicit_post_bootstrap_depth_ = explicit_bootstrap_lvl;
+                } else {
+                    LOG_AND_THROW_STREAM("Internal error: explicit_bootstrap_lvl is < 0: " << explicit_bootstrap_lvl);
+                }
+            } else {
+                // we have previously set the explicit_post_bootstrap_depth_; make sure we get the same value.
+                if (explicit_post_bootstrap_depth_ != explicit_bootstrap_lvl) {
+                    LOG_AND_THROW_STREAM("Circuit error: explicit_post_bootstrap_depth_ is was previously set to "
+                                         << explicit_post_bootstrap_depth_ << ", but now is "
+                                         << explicit_bootstrap_lvl);
+                }
+            }
```

Comment:
This method may need more comments to explain how the depth is set.

Reply:
Added extensive comments.